### ISSUE TITLE
[UX] Improve comment settings on both node form and node type form

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#764]
Not all users get the difference between the two. How about some help text there. This is from [Administering a content type's comment settings](https://www.drupal.org/node/777818):

>- Hidden: No comments are allowed, and past comments are hidden.
>- Closed: No comments are allowed, but any past comments remain visible.
>- Open: Any future content of this type is open to new comments.

**Comment by docwilmot** [comment]
+1 this. I still have to stop and think.

**Comment by klonos** [comment]
We could even shorten this a bit (we can skip help text for "open" since it is self-explanatory):

>- Hidden: No comments allowed, past comments are hidden.
>- Closed: No comments allowed, but any past comments remain visible.

**Comment by quicksketch** [comment]
I think @jenlampton has a long-standing issue on this. It doesn't make any sense for the *default* setting of a node to be *either* hidden or closed. They mean the same thing for new pieces of content, so it doesn't make any sense to choose between the two of them. Comments should either be open or closed by default. "Hidden" is an option that should only exist after comments have actually been made on an individual node.

**Comment by docwilmot** [comment]
> It doesn't make any sense for the default setting of a node to be either hidden or closed.

very true too. Seems the open/closed should be on the node type and the option to hide should be on the node. Although then we could simplify even further and only show the option to hide if the node actually has a comment.
* Open coments
* Close comments
  - [ ] Hide existing

**Comment by klonos** [comment]
Fair enough, but if we moved the "hidden" option on the node level, how would one apply this setting to a content type level (to all existing nodes of a specific content type). Shouldn't we provide a VBO option to set this?

**Comment by jenlampton** [comment]
@klonos, what you are asking for is a feature request (it doesn't exist now, nor has it ever existed in Drupal) but yes - I think it's a good one. Can you please open a new issue?

The option you see on the content type form only sets the default values for new nodes of that type.

@docwimot I love your suggestion for the checkbox. - Far less confusing!

Here's the issue on drupal.org https://www.drupal.org/node/989452

**Comment by quicksketch** [comment]
Great start on the PR! But it needs a small amount of work to ensure existing node types get updated to be only open or closed (and not hidden). https://github.com/backdrop/backdrop/pull/748#issuecomment-75502865

**Comment by quicksketch** [comment]
I love the automatically present or hidden checkbox for hiding existing comments. However, I found a small JS bug in that you can set comments to be Open after checking the "Hide existing comments" checkbox, then the vertical tabs show conflicting settings:

![comment-settings](https://cloud.githubusercontent.com/assets/100206/6324557/5de7ca7c-baf0-11e4-8c9b-988421845993.png)

I know the other issue was held up by a bug in vertical tabs (it's the thing adding that trailing space after the word "Open"), that does not need to be fixed in this PR, as it's a separate problem (but I don't think we have an issue for it yet).

**Comment by klonos** [comment]
> ...I don't think we have an issue for it yet.

Here you go!

**Comment by klonos** [comment]
> So, what you are asking for is a feature request (it doesn't exist now, nor has it ever existed in Drupal) but yes - I think it's a good one. :)

> Can you please open a new issue?

I personally see it as a feature regression because previously one could hide all comments for content of a specific node type. With the change we are about to make, achieving the same thing will be a tedious job. New issue filed.

**Comment by jenlampton** [comment]
> I personally see it as a feature regression because previously one could hide all comments for content of a specific node type.

That's not actually true.  Before this change you could set the default behavior - that means you could set the comment type for all NEW content of a type. But changing the default behavior (now and previously) will not change this setting on any existing piece of content.

**Comment by klonos** [comment]
You are absolutely right Jen. I take my comment about the regression back.

...actually I have to admit that I am a bit embarrassed. I've been a Drupalist for 5 and a half years and I always thought that this setting was a "global" setting. My eyes scanned too fast through that form I guess and somehow chose to ignore the "new" part in the label (Default comment setting for *new* content). I guess the fact that I never gave the "Administer comments and comment settings" permission to users other than the admin played a part in this. It simply was not needed in the sites I've built so far so this workflow became a habit making me take things for granted.

**Comment by jenlampton** [comment]
just pushed some updates to https://github.com/backdrop/backdrop/pull/748 that solve the issues pointed out in [nate's comment](https://github.com/backdrop/backdrop-issues/issues/764#issuecomment-75503512)

**Comment by klonos** [comment]
...shamelessly promoting/bumping this one :smiling_imp: ...or how we used to do it in d.org pre-subscribe button era: Subscribing!!!

**Comment by klonos** [comment]
So to get this right, on the content type form we'll have "enabled"/"disabled" and on node edit/create form we'll have "open"/"closed" (plus a checkbox for hiding past comments). Right?

**Comment by sutibun** [comment]
Subscribing :p

**Comment by quicksketch** [comment]
Yay, this PR looks great. Code is pretty much perfect IMO.

> So to get this right, on the content type form we'll have "enabled"/"disabled" and on node edit/create form we'll have "open"/"closed" (plus a checkbox for hiding past comments). Right?

That's right.

This PR doesn't contain any tests however :frowning:. I guess "hidden" functionality had never been tested in the past? It would be good to expand test coverage to include it.

**Comment by quicksketch** [comment]
I'm working on adding tests for this item so we can get it merged.

**Comment by quicksketch** [comment]
New PR added at https://github.com/backdrop/backdrop/pull/1149. @jenlampton's code looks perfect to me, so the only addition is tests to check this new functionality.

While working on the comment.test file, I switched the comment tests base class to use the "testing" profile to help speed up tests further. This skips installing modules like update, views, field ui, etc. when they're not needed for testing.

**Comment by quicksketch** [comment]
I've merged in https://github.com/backdrop/backdrop/pull/1149 that includes the new tests. Thanks all!

**Comment by klonos** [comment]
Thank you so much. Another small step towards perfection!

**Comment by klonos** [comment]
...a few minor things. Re-open, or file follow-ups?

**Comment by klonos** [comment]
...a list so I don't forget them:

- Shouldn't it be "Close**d** comments" instead of "Close comments"? (content type edit form.
- Shouldn't the "Comments per page" and the rest of the comment settings be hidden if comments are set to closed? ...and also respective par of the vtab summary.
- Space before comma in the vtab summary.

**Comment by quicksketch** [comment]
Let's handle them here.

**Comment by quicksketch** [comment]
> Shouldn't it be "Closed comments" instead of "Close comments"? (content type edit form).

Agreed.

> Shouldn't the "Comments per page" and the rest of the comment settings be hidden if comments are set to closed? ...and also respective part of the vtab summary.

Hm, actually I don't think so. Because comments can still be enabled per-node. The Open/Closed setting is only the *default*, so number of comments per page and other settings are still relevant if comments are opened on the node individually. However, I think this brings up a great point that you should be able to truly *disable* comments entirely on a single type, rather than it only being a default.

> Space before comma in the vtab summary.

Sure, an existing issue but we can fix it.

**Comment by quicksketch** [comment]
Minor fixes PR addresses 1 and 3 but not 2 from @klonos's list: https://github.com/backdrop/backdrop/pull/1152

**Comment by klonos** [comment]
> ...comments can still be enabled per-node. The Open/Closed setting is only the *default*, ...you should be able to truly *disable* comments entirely on a single type, rather than it only being a default.

Yes, my point exactly. I admit that despite being a Drupal old-timer, I always was under the impression that this was what this setting actually did (disabling comments for that content type permanently without the option to enable on a per-node basis). I always missed the word "default" in the radio set label somehow. Shame on me!
